pkg/operator/controller/dns_status.go

```diff
@@ -114,13 +108,16 @@ func computeDNSDegradedCondition(oldCondition *operatorv1.OperatorCondition, clu
 			status = operatorv1.ConditionTrue
 			degradedReasons = append(degradedReasons, "NoDNSPodsAvailable")
 			messages = append(messages, "No DNS pods are available.")
-		case intstrErr != nil:
-			degradedReasons = append(degradedReasons, "InvalidDNSMaxUnavailable")
-			messages = append(messages, fmt.Sprintf("The DNS daemonset has an invalid MaxUnavailable value: %v", intstrErr))
-		case int(numberUnavailable) > maxUnavailable:
-			status = operatorv1.ConditionTrue
-			degradedReasons = append(degradedReasons, "MaxUnavailableDNSPodsExceeded")
-			messages = append(messages, fmt.Sprintf("Too many DNS pods are unavailable (%d > %d max unavailable).", numberUnavailable, maxUnavailable))
+		case want > have:
+			// Mark as degraded only if it's been determined that it's not already degraded as a recent last transition.
+			if oldCondition != nil && oldCondition.Status == operatorv1.ConditionFalse {
```

Comment:
Is this redundant code? It will already [exit](https://github.com/openshift/cluster-dns-operator/pull/373/files#diff-32495132facf7e0819a407af732514958b198d9e40236656bc43b093345d1539R127) here if the condition is already degraded and the transition time is recent.

Am I missing something?

Reply:
If it doesn't add the degradedReason here (or elsewhere) then it doesn't get to L127.  But it could add the degradedReason for other conditions, like want=0 or have=0, which we do want to check in L127.  Does that answer your question?

Reply:
Hmm. I see the difference I think you are referring to, but it's not super clear to me of the impact. I'll keep studying.

Reply:
Okay, I'm still a little confused on what this is accomplishing exactly, so I'll take it line-by-line:

```
if oldCondition != nil && oldCondition.Status == operatorv1.ConditionFalse {
```
- This says you will **only** add `NotAllDNSPodsAvailable` condition if the previous condition was `false`. If it was `true`, it won't do anything.
- To be clear, this logic doesn't prevent status flapping (false -> true, true -> false), but instead prevents changing of the status reason and message while the status is already `true`.
- Sounds like this is adding a hurdle for the logic to set `NotAllDNSPodsAvailable`: the condition must go false before it becomes true again.
- Your goal was to reduce the time in degraded conditions right? But if the status condition was already `true`, it'd just be a reason / message / last transition time update, which if I understand correctly, seems not a bad thing to do.
    - Would allowing the reason / message to update while conditions are still `True` be beneficial information to the user?

```
timeSinceTransition := currentTime.Sub(oldCondition.LastTransitionTime.Time)
if timeSinceTransition > transitionUnchangedToleration {
```
- This says you will **only** add `NotAllDNSPodsAvailable` if the previous condition was older than `transitionUnchangedToleration`.
- I understand this logic a bit better. However, I'm still struggling to see how it isn't redundant with [L127](https://github.com/openshift/cluster-dns-operator/pull/373/files#diff-32495132facf7e0819a407af732514958b198d9e40236656bc43b093345d1539R127).
   - i.e. if you remove the check `if timeSinceTransition > transitionUnchangedToleration {`, then `lastTransitionTimeIsRecent` is going to unset the status by `return *oldCondition` in L128.
   - To confirm this, I ran the [unit test](https://github.com/openshift/cluster-dns-operator/pull/373/files#diff-02fb4bb598cb2a9cd3ce8f83f8c21b817f62b8083800f2e613990e294b0d876aR584), removed the `if timeSinceTransition > transitionUnchangedToleration {` logic, and the result was the same, it still returns `Degraded=ConditionFalse` because L127 catches it.

So, my apologies if I don't totally understand, but it feels like the check to make sure the condition is `False` before adding the `NotAllDNSPodsAvailable` condition is a bit of a hurdle that tries to slow down status changes while the check for `transitionUnchangedToleration` seems redundant.

Reply:
Thanks for the feedback.  You're right, it's not quite right yet, especially for the case where it comes into this function as already Degraded.  I will mark this as WIP for now.